### DISCUSSION TITLE
Translating XML to Capn Proto (and back again)

### DISCUSSION THREAD

**Discussion by dontlaugh** [capnproto/capnproto#2137] [General]
I was having an idle daydream: can I reliably represent XMPP schemas as Capn Proto schemas?

XMPP streams are kind of weird in that they are an unbounded, streaming XML "document" end-to-end, but the messages in the middle are reasonably well-specified. Whether the protocol can be translated to Capn Proto is interesting, but I'm just concerned with message types for now.

https://xmpp.org/extensions/

Here is an example of a typical message/packet/object you might encounter. A [Multi User Chat (MUC) Token Invite](https://xmpp.org/extensions/xep-0488.html).

```xml
<iq type='result' from='news@commons.example.org' to='louise@example.org' id='request2'>
  <token xmlns='urn:xmpp:muc-token-invite:0'
         delay='604800'
         counter='5'
         >lyQZ1RzacYTlf3svGODYq1xVabNnMc2x</token>
</iq>
```

Here is another, a query for a server's roster

```xml
<iq type="get" id="roster_1" from="user1@domain.com" to="domain.com">
  <query xmlns="jabber:iq:roster"/>
</iq>
```

In XMPP, you have these common envelope types, and then things get increasingly more _eXtensible_ in the inner parts of the tree. Implementations should adhere to a minimum standard of conformity, and can ignore certain extensions. Implementations can tell each other about extensions, too.

But really, at first, I'm concerned with serialization. If I can't reliably use the standard Capn Proto code generators to serialize these objects, it's a no go.

For those familiar with XML, XMPP, and Capn Proto: do you think Capn Proto's schema language could represent something like this? I'm new to Capn Proto.

```
struct IQ (Inner) {
  type @0 :Text;
  id @1 :Text;
  from @2 :Text;
  to @3 :Text;
  inner @4 :Inner;   # could be a LOT of different stuff
}
```

**Comment by kentonv** [comment]
You could certainly write an `XmlCodec` in the style of the existing `JsonCodec` to do something like this. Of course, the structure of XML doesn't really map 1:1 with structs-and-lists style of Cap'n Proto and JSON, so the problem is more complex. And I suppose you'd probably want to use a real XML parser library under the hood rather than roll one from scratch like `JsonCodec` does.
